### PR TITLE
PCHR-2032: Share Leave Requests component

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

The Leave Requests component is now a shared component in L&A. It is now used by both managers and admins in their dashboards.

## After

new

## Technical Details

1. Leave Requests component moved to the shared folder
2. The component was injected to admin dashboard
3. Styles for the component were moved to L&A (also see https://github.com/compucorp/civihr-employee-portal-theme/pull/183)
4. Tests were moved and amended accordingly

## Comments

1. Gulp watch / tests running is fixed, now it supports `*.spec.js` files only
2. All L&A test files were renamed to `*.spec.js`
3. Both admin and manager layouts are developed according to the layout spec:

Admin:
image

Manager SSP:
image

Admin SSP:
image

---

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js [hunk 1]
```diff
@@ -1,27 +1,31 @@
...
-      'OptionGroup', 'dialog', 'shared-settings', controller]
+    controllerAs: 'vm',
+    controller: [
+      '$controller', '$log', '$q', '$rootScope', 'shared-settings', 'checkPermissions',
+      'Contact', 'AbsencePeriod', 'AbsenceType', 'LeaveRequest', 'OptionGroup',
+      'dialog', controller]
   });
 
-  function controller ($log, $q, $rootScope, Contact, AbsencePeriod, AbsenceType, LeaveRequest, OptionGroup, dialog, sharedSettings) {
+  function controller ($controller, $log, $q, $rootScope, sharedSettings, checkPermissions, Contact, AbsencePeriod, AbsenceType, LeaveRequest, OptionGroup, dialog) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Remove `$controller` as its not being used.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js [hunk 2]
```diff
@@ -53,6 +57,7 @@ define([
     vm.isFilterExpanded = false;
     // leaveRequests.table - to handle table data
     // leaveRequests.filter - to handle left nav filter data
+    vm.isAdmin = false; // this property is updated on controller initialization
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]:
```
// leaveRequests.table - to handle table data
// leaveRequests.filter - to handle left nav filter data
```
These comments are meant for `vm.leaveRequests`, please keep that order. Also alphabetically `isAdmin` should go before `isFilterExpanded`.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/manage-leave-requests.spec.js [hunk 3]
```diff
@@ -441,11 +448,19 @@
                 controller.refresh();
               });
 
-              it('Contact API gets called with region value', function () {
-                expect(Contact.leaveManagees).toHaveBeenCalledWith(controller.contactId, jasmine.objectContaining({
-                  region: mockRegion
-                }));
-              });
+              if (isAdmin) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/manage-leave-requests.spec.js]: 1. In the current setup I dont think it is even testing both the scenarios, because you are not changing the value of isAdmin.
2. Having if-else blocks in the unit test should be avoided.
3. Multiple test cases must be test up to test different scenarios
```javascript
describe('when admin', function() {
 beforeEach(function(){
   // setAdmin();
 })
  // Test every case where admin needs to be true in this describe
})
describe('when not admin', function() {
 beforeEach(function(){
   // setManager or staff();
 })
  // Test every case where admin needs to be false in this describe
})
```
Please check all such scenarios

**Comment by deb1990** [comment]
Also write backstop tests.

**Comment by deb1990** [comment]
5_1_admin_leave_dashboard
The layout is bit different in Admin. Please check

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js (hunk 1 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js (hunk 2 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Done.

**Comment by igorpavlov** [comment]
@deb1990 

> The layout is bit different in Admin. Please check

This is not the layout I see:

image

**Comment by igorpavlov** [comment]
@deb1990 

> The layout is bit different in Admin. Please check

Show it be also applied to Manager's layout of only Admin?

File: uk.co.compucorp.civicrm.hrleaveandabsences/gulpfile.js [hunk 4]
```diff
@@ -103,35 +103,24 @@ var test = (function () {
     /**
      * Runs the tests for a specific source file
      *
-     * Looks for a test file (*_test.js or *.spec.js) in `test/`, using the same path
+     * Looks for a test file (*.spec.js) in `test/`, using the same path
      * of the source file in `src/leave-absences/`
-     *   i.e. src/leave-absences/models/model.js -> test/models/model_test.js
-     *   or   src/leave-absences/models/model.js -> test/models/model.spec.js
-     *   @TODO test files *_test.js should be renamed to *.spec.js
+     *   i.e. src/leave-absences/models/model.js -> test/models/model.spec.js
```
Comment (bmnepali) [inline uk.co.compucorp.civicrm.hrleaveandabsences/gulpfile.js]: Seems there is a wrong indentation here.

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/admin-dashboard-leave-requests.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      casper.then(function () {
+        // Wait for the navigation to load
+        casper.waitUntilVisible('.chr_admin-dashboard_side-nav', function () {
+          // Select the Leave Requests tab
+          casper.click('.chr_admin-dashboard_side-nav [ui-sref=requests]');
+          // Wait until the header of the Leave Requests component is loaded
+          casper.waitUntilVisible('.chr_manage_leave_requests__header', function () {
+            /*
+             * @TODO FIX and improve this Backstop script
+             * Something unexpected happens with Casper here.
```
Comment (bmnepali) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/admin-dashboard-leave-requests.js]: > Something unexpected happens with Casper here.

Do you have any exceptions or errors here?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js [hunk 6]
```diff
@@ -50,14 +50,20 @@ define([
...
       }
     };
     vm.filteredUsers = [];
+    vm.filtersByAssignee = [
+      { type: 'all', label: 'All' },
+      { type: 'unassigned', label: 'Unassigned' },
+      { type: 'me', label: 'Assigned To Me' }
+    ];
     vm.isFilterExpanded = false;
+    vm.isAdmin = false; // this property is updated on controller initialization
```
Comment (bmnepali) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Please set the one-line assignments before  multiline assignments:
```js
vm.isFilterExpanded = false;
vm.isAdmin = false;

vm.filtersByAssignee = [
    { type: 'all', label: 'All' },
    { type: 'unassigned', label: 'Unassigned' },
    { type: 'me', label: 'Assigned To Me' }
];

...
```

Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: I am not sure about that as if you look at https://compucorp.atlassian.net/wiki/display/PCHR/Angular+1+Style+Guide+discussion there is a rule about the var types and the alphabetical order. However, I am not sure I see a rule for multiline assignments.

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/admin-dashboard-leave-requests.js (hunk 5 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/admin-dashboard-leave-requests.js]: Added a comment.

File: uk.co.compucorp.civicrm.hrleaveandabsences/gulpfile.js (hunk 4 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/gulpfile.js]: Fixed.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js (hunk 6 above)
Comment (bmnepali) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Right! However, it's better to keep the assignments grouped appropriately as I suggested . What's your opinion on this?

Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: @AkA84 could you please help us here?

Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: always people group single line assignments together and then multiline assignments together. you don't have to put an empty line between the two

Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Fixed.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js [hunk 7]
```diff
@@ -427,15 +453,16 @@ define([
...
         status_id: prepareStatusFilter(filterByStatus),
         type_id: filters.selectedAbsenceTypes ? filters.selectedAbsenceTypes.id : null,
         from_date: {
           from: filters.selectedPeriod.start_date
         },
         to_date: {
           to: filters.selectedPeriod.end_date
-        }
+        },
+        unassigned: filters.assignedTo === 'unassigned' ? 1 : null
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Just `filters.assignedTo === 'unassigned'` should be enough.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js [hunk 8]
```diff
@@ -427,15 +453,16 @@ define([
 
       return {
         contact_id: prepareContactID(),
-        managed_by: vm.contactId,
+        managed_by: vm.isAdmin && filters.assignedTo !== 'me' ? undefined : vm.contactId,
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Add a parentheses around the condition, easier to read, `(vm.isAdmin && filters.assignedTo !== 'me')`

**Comment by deb1990** [comment]
manager_leave___hr17-1
@igorpavlov Are the UI changes in SSP approved by Mike?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js (hunk 8 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js (hunk 7 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/manage-leave-requests.js]: Done.

**Comment by igorpavlov** [comment]
@deb1990 Mike recommended to revert the layout for the manager back, so I had to refactor the HTML/CSS code to support both. Please see screenshots in the comment section.

**Comment by deb1990** [comment]
@igorpavlov The Styles for the ALL/USASSIGNED.. button is not looking good. I understand that it might not be possible to make it exactly similar to admin as both pages uses different themes. So I suggest to import the SCSS files from shoreditch into SSP so that the `outlined` classes are available in SSP too. 
cc @AkA84

**Comment by igorpavlov** [comment]
@deb1990 as per styles you are absolutely right. Please see the updated screenshots in the PR. However, I decided *not* to import the button styles. Instead, I added some patching CSS with an appropriate comment. The reason is to incapsulate the patch to eliminate any possibility of breaking styles in any other places for SSP.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss [hunk 9]
```diff
@@ -57,20 +84,68 @@
...
-  .panel {
-    background-image: none !important;
+  .chr_manage_leave_requests__sub-header .checkbox {
+    padding-top: 5px;
   }
 
-  .panel-body, .chr_vertical_tabs {
-    background-color: #f4f7f8;
+  .btn-secondary-outline {
+    color: #4D5663;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: We should not create class for this, because we are defining a new variant for button. So it be not be under `chr_manage_leave_requests`, these outlined classes might need to be other places in SSP. So if we cant import from shoreditch, we should define these classes under "https://github.com/compucorp/civihr-employee-portal-theme/blob/bc9c947f882906969d71b2cefd4f126e4d721fcf/civihr_default_theme/scss/components/_button.scss#L9-L8" using the variables for `btn-primary-outline`, `btn-secondary-outline`, `btn-info-outline`, `btn-success-outline`,  `.btn-warning-outline` and `.btn-danger-outline`

Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: shouldn't we be able to import them from shoreditch?

Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: The shoreditch files also contains other styles for buttons, it might break SSP. But we can separate the outlined part and import.

Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Sounds like an overkill to me. My idea was to incapsulate any changes in SSP to this component, that's exactly what I made.

Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: @igorpavlov I had a discussion with @AkA84 . Here is the outcome. We should import all the outline classes from shoreditch to SSP. But we cant do that directly because `org.civicrm.shoreditch/scss/bootstrap/overrides/style/_buttons.scss` also contains the some styles which are not related to outlined buttons. So we need to take the code related to only outlined classes and move that to a new folder(may be called "variant/extension") . Then we can import that file directly to SSP without worrying about breaking anything. 
Also this would be a better structure in shoreditch, because right now in overrides folder, we are also including the new variants (like outlined buttons), but it should go to a new folder.

So please create a new folder under `scss/bootstrap/extensions/_outlined-buttons.scss`. And import the same from SSP.
Note: the Shoreditch changes should go to 1.6 and then synced.

Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: @deb1990 @AkA84 ok, will do.

Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss [hunk 10]
```diff
@@ -57,20 +84,52 @@
  * @TODO remove this patch when Shoreditch theme is migrated to SSP
  */
 .chr_manage_leave_requests {
-  .btn.ui-select-toggle {
-    padding: 4px 10px !important;
-    text-transform: none !important;
+  &[data-role="admin"] {
+    .ui-select-toggle {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Do we really need to override `ui-select-toggle`? In my last PR I fixed the styles in shoreditch. What are you getting if you dont have this override?

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss [hunk 11]
```diff
@@ -57,20 +84,52 @@
...
+      text-transform: none !important;
+      box-shadow: none !important;
+    }
+
+    .form-group {
+      overflow: inherit !important;
+    }
+
+    .panel {
+      background-image: none !important;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Why is it needed? Does panel have a background image set?

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss [hunk 12]
```diff
@@ -57,20 +84,52 @@
...
+    }
+
+    .panel-body, .chr_vertical_tabs {
+      background-color: #f4f7f8;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Use a variable here. This color is used in manager-calendar also.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss [hunk 13]
```diff
@@ -57,20 +84,52 @@
...
+    }
+
+    input[type="search"] {
+      width: 100% !important;
+    }
+
+    .ui-select-toggle .btn-link i {
+      margin-top: 6px;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Why is this required?

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss [hunk 14]
```diff
@@ -57,20 +84,52 @@
...
+    }
   }
 
-  .form-group {
-    overflow: inherit !important;
+  .btn-link {
+    border: none !important;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Why is this required? `btn-link` does not have border and backgound color by default.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss [hunk 15]
```diff
@@ -57,20 +84,52 @@
...
-  .panel-body, .chr_vertical_tabs {
-    background-color: #f4f7f8;
+/*
+ * This is a temporary patch which fixes CSS in the component (for SSP only)
+ * @TODO remove this patch when Shoreditch theme is migrated to SSP
+ */
+.page-manager-leave {
+  .chr_manage_leave_requests {
+    .btn-sm {
+      font-size: 100% !important;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Why is this required?

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 10 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: @deb1990 

image

It is fine in terms of layout, but it has a shadow which looks a bit inconsistent. Mine CSS rule removes the shadow. Shall I leave just a shadow rule or shall I remove rules for this selector completely?

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 11 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: screen shot 2017-07-11 at 10 14 29

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 10 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Please remove the selector altogether. And the `box-shadow`is a theme problem, so if we need to fix it we should do it only in shoreditch. But I would say that its not a big issue(fixing in Shoreditch) and can be skipped for now.

Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 12 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 13 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: image

The X looks bad to me without this rule.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 10 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Actually, not only this. In SSP it is off:

image

Can I apply a additional rule to SSP only?

```css
.ui-select-toggle {
  padding: 7px 12px !important;
  font-size: 100% !important;
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 13 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: It looks fine in SSP, 
manager_leave___hr17-1
Please check why exactly it is broken in admin

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 14 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Removed as it is fixed in latest Shoreditch.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 15 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: > .chr_manage_leave_requests__sub-header .checkbox

Otherwise not well aligned vertically and looks odd.

> .btn-sm

Otherwise in SSP looks smaller than other buttons, however in Admin (which has a priority because it uses Shoreditch) we must use btn-small according to the Invision and the Styleguide.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss (hunk 13 above)
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_manage-leave-requests.scss]: Yes, you are right. After syncing all CSS with Shoreditch all problems are gone.

**Comment by igorpavlov** [comment]
Squashed PR Feedback commits into one.
